kernel/fork.c

```diff
@@ -107,11 +107,11 @@ int do_fork(bool vfork)
    ASSERT(curr->state_regs != NULL);
 
    child->state_regs--; // make room for a regs_t struct in child's stack
-   *child->state_regs = *curr->state_regs; // copy parent's regs_t
+   *child->state_regs = *curr->user_state_regs; // copy parent's regs_t
```

Comment:
Here the preemption is disabled. So, at most an IRQ can happen. There are 3 cases:
1. An IRQ happens at _before_ this statement: that's not a problem because the stack will change and eventually the IRQ handler will return and everything will be correct as before.
2. An IRQ happens _after_ this statement: that's not a problem because we already copied the registers.
3. An IRQ happens _during_ this statement: **yep, that's a problem**. But that doesn't mean we need a separate `user_state_regs` field etc. The bug I see is simply that this copy operation is not IRQ safe. Therefore, the simplest thing we can do to fix this is to perform the copy while keeping the interrupts disabled:

```C
disable_interrupts_forced();
{
   *child->state_regs = *curr->state_regs;
}
enable_interrupts_forced();
```
Does this make sense or am I missing something else?
My general thinking is to avoid introducing "new state" in any form unless that's unavoidable. A bigger state means almost always "more complexity" and makes the whole thing harder to reason about. Of course, if you show me that my suggestion doesn't work and somehow IRQs before that point somehow change the state in an undesirable way, can talk about that.

The assumption that an IRQ restores the state as it was before is of **critical importance**. If that has a bug, let's fix it as well.

Reply:
Because of the above, simply turning off the interrupt when the `*child->state_regs = *curr->state_regs;`statement is executed won't solve the problem.  So the user's regs state must be saved somewhere else

Reply:
@aenrbes Thank you for clarifying that. It wasn't obvious and I didn't spent enough time on the problem. 
So, in my understanding now what happens is that after that initial `enable_preemption()` in `do_syscall()`, the current task gets preempted (the IRQ 0 triggering that is irrelevant, the code could also have yield etc.). When that happens, `state_regs` points to the current _kernel_ state which is later used as the regs for the child, while `running_in_kernel=false` for the child task and that leads to any sort of fancy corruption during the task switch. While with your fix it works because we save the USER regs separately and make the new task work with that. Is that correct?

Now, if that's the case, my follow-up question for you is: will `user_state_regs` used only by `fork()` or it should play a role elsewhere?

* If it's only used by fork(), we could make fork() be a special syscall and be called by `do_special_syscall()`. That will allow us to say that we don't want preemption enabled by default for it. Then, since it's sub-optimal to introduce such a large non-preemptable code, we could add another abstraction to the syscall infra with a new flag like SYSFL_RAW_REGS and make sys_fork() and sys_vfork() to accept a `regs_t *` parameter. Then, the parameter will be forwarded to do_fork() and saved before enabling the preemption. The gain with this approach is avoiding introducing new per-task state and is used only temporarily by one syscall. [ If you don't wanna do all of that, you could simply flag `fork` and `vfork` as SYSFL_NO_PREEMPT and check if that solves the problem as it should in my current understanding. I'll do the follow-up optimization, if you don't want/have time ].

* On the other side, if there is another use case for that or if creating new use-cases for the new field will simplify something, then it makes sense add the new state. I'm not so sure about that and haven't spent enough time thinking about that as I'm out of home most of the time these days. What do you think?

I'm wondering at this point if something can go wrong with the preemption of user space tasks. In theory, it's all good because the new kernel stack and state are added on the top of the user space one:
```

+---------------+
|  IRQ handler  |
+---------------+
|  kernel funcs |     <- state_regs (kernel)
+---------------+
|  syscall      |
+---------------+
|  user regs    |    <- state_regs (user space)
+---------------+
```

That's the model I used for a long time after the "early days" when I kept a copy of the state registers explicitly in the struct task. That was pointless since they have been already saved on the stack. The new model allowed a long series of simplifications and performance improvements but hey.. I missed the fork() case. Do you see some other limitation to that model besides the specific fork() case?

Reply:
You're absolutely right. user_state_regs is used only by fork(). In fact, the first solution I came up with was to pass the user_regs pointer as a function argument, just because it changes the code a little bit more.

And I don't see any other problems with the model of preemption of user space tasks at the moment.

I have modified the code using the function parameter method, tested it, and committed it below

Reply:
Thanks for re-working your change, @aenrbes. It looks great now.
I'm going to merge it.